pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -55,268 +133,487 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
-	podInformer cache.SharedInformer
+func (r *ReconcileNodeMaintenance) adjustLeaseTimeIfRequired(specAnnoData *ReconcileNodeMaintenanceSpecInfo, statusAnnoData *ReconcileNodeMaintenanceStatusInfo, lease *coordv1beta1.Lease) error {
+	duration := statusAnnoData.OpDeadline.Sub(statusAnnoData.OpStart)
+	durationSeconds := int64(duration.Nanoseconds()) / int64(time.Second)
+
+	diffTime := int64(specAnnoData.RequestedMaintenanceWindowSeconds) - durationSeconds
+
+	if diffTime < -3 || diffTime > 3 {
+		r.reqLogger.Infof("Lease time has been changed. %d", int(durationSeconds))
+
+		specSeconds := int32(durationSeconds)
+		lease.Spec.LeaseDurationSeconds = &specSeconds
+
+		if err := r.client.Update(context.TODO(), lease); err != nil {
+
+			return err
+		}
+	}
+	return nil
 }
 
-var Handler ReconcileHandler
+func (r *ReconcileNodeMaintenance) processMaintModeOn(statusAnnoData *ReconcileNodeMaintenanceStatusInfo, node *corev1.Node) (reconcile.Result, error) {
 
-// Reconcile reads that state of the cluster for a NodeMaintenance object and makes changes based on the state read
-// and what is in the NodeMaintenance.Spec
-// Note:
-// The Controller will requeue the Request to be processed again if the returned error is non-nil or
-// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
-func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	reqLogger := log.WithFields(log.Fields{"Request.Namespace": request.Namespace, "Request.Name": request.Name})
-	reqLogger.Info("Reconciling NodeMaintenance")
+	nodeName := node.ObjectMeta.Name
 
-	// Fetch the NodeMaintenance instance
-	instance := &kubevirtv1alpha1.NodeMaintenance{}
-	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
-	if err != nil {
-		if errors.IsNotFound(err) {
-			// Request object not found, could have been deleted after reconcile request.
-			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
-			// Return and don't requeue
-			reqLogger.Infof("NodeMaintenance Object: %s Deleted ", request.NamespacedName)
-			return reconcile.Result{}, nil
+	switch statusAnnoData.Status {
+	case NodeStateInit:
+		{
+			statusAnnoData.Status = NodeStateCordon
+			if _, err := r.setStatusAnnotation(node, statusAnnoData); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+		fallthrough
+
+	case NodeStateCordon:
+		{
+			if drainer, kclient, err := r.initDrainer(); err == nil {
+				if err := r.runCordonOrUncordon(drainer, kclient, nodeName, true); err != nil {
+					// TODO: is that the right thing? or should it rather report error.
+					return reconcile.Result{Requeue: true, RequeueAfter: RequeuDrainingWaitTime * time.Second}, nil
+				}
+			} else {
+				return reconcile.Result{}, err
+			}
+			statusAnnoData.Status = NodeStateEvicting
+			_, err := r.setStatusAnnotation(node, statusAnnoData)
+			if err != nil {
+				return reconcile.Result{}, err
```

Comment:
This demonstrates the fragility of relying on `statusAnnoData.Status`

If ``setStatusAnnotation()`` fails, it does not mean that we are in state `NodeStateCordon`, nor that cordon needs to be reattempted, nor that eviction should be delayed - but using `statusAnnoData.Status` in this way asserts that all these things are true anyway.

It would be better to check the actual condition rather than a flag that may be out of sync with reality.
In this specific case:  
````
nodeCorodoned = ...
if !nodeCordoned  { r.runCordonOrUncordon() ... }
````

I'm not saying to delete `statusAnnoData.Status`, but let it just be for providing information to admins rather than having the controller make decisions based on its value.

Reply:
Ok, I would keep the state of the request somewhere. That leads us to the same problem as with keeping the drainer per reconcile request:  
I can't just keep it in the ReconcileNodeMaintenance structure - we have one such object per control loop; therefore we would need to add a map from Request object to some structure that holds the status (and drainer) and delete that one when the request is completed. Do you approve of this course of action?

Reply:
I can't extend the reconcile.Request structure either - it is created inernally and I can't hook the creation of the request.

Reply:
> Ok, I would have keep the state of the request somewhere. 

Yes, keep it somewhere, in the same place it is now BUT only to provide information to the admin, NOT to use as part of the state machine logic.

> That leads us to the same problem as with keeping the drainer per reconcile request:
> I can't just keep the state of the reconcile request in the ReconcileNodeMaintenance structure - we have one such object per control loop; therefore we would need to add a map from Request object to some structure that holds the status (and drainer) and delete that one when the request is completed. Do you approve of this course of action?

No. See above

Reply:
done